neps/nep-0508.md

```diff
@@ -0,0 +1,151 @@
+
+## Specification
+
+### High level assumptions
+
+* Some form of State sync (centralized or decentralized) is enabled.
+* Flat state is enabled.
+* Shard split boundary is predetermined. In other words, necessity of shard splitting is manually decided.
+* Merkle Patricia Trie is undelying data structure for the protocol.
+* Minimal epoch gap between two resharding events is X.
+
+### High level requirements
+
+* Resharding should work even when validators stop tracking all shards.
+* Resharding should work after stateless validation is enabled.
+* Resharding should be fast enough so that both state sync and resharding can happen within one epoch.
+* Resharding should not require additional hardware from nodes.
+* Resharding should be fault tolerant
+  * Chain must not stall in case of resharding failure.
+  * A validator should be able to recover in case they go offline during resharding.
+* No transaction or receipt should be lost during resharding.
+* Resharding should work regardless of number of existing shards.
+
+### Out of scope
+
+* Dynamic resharding
+* Shard determination logic (shard boundary is still determined by string value)
+* TBD
+
+### Required protocol changes
+
+TBD. e.g. configuration changes we have to introduce
+
+### Required state changes
+
+TBD. e.g. additional/updated data a node has to maintain
+
+### Resharding flow
+
+TBD. how resharding happens at the high level
+
+## Reference Implementation
+
+TBD
+
+## Security Implications
+
+[Explicitly outline any security concerns in relation to the NEP, and potential ways to resolve or mitigate them. At the very least, well-known relevant threats must be covered, e.g. person-in-the-middle, double-spend, XSS, CSRF, etc.]
+
+## Alternatives
+
+* Why is this design the best in the space of possible designs?
+  * TBD
+* What other designs have been considered and what is the rationale for not choosing them?
+  * TBD
+* What is the impact of not doing this?
+  * TBD
+
+## Future possibilities
+
+As noted above, dynamic resharding is out of scope for this NEP and should be implemented in the future. Dynamic resharding includes the following but not limited to:
+* automatic determination of split boundary
+* automatic shard splitting and merging based on traffic
+
+Other useful features that can be considered as a follow up:
+* account colocation for low latency across account call
+* shard on demand
+
+## Consequences
+
+### Positive
+
+* Workload across shards will be more evenly distributed.
+* Required space to maintain state (either in memory or in persistent disk) will be smaller.
+* State sync overhead will be smaller.
+* TBD
+
+### Neutral
+
+* Number of shards is expected to increase.
+* Underlying trie structure and data structure are not going to change.
+* Resharding will create dependency on flat storage and state sync.
```

Comment:
I wouldn't say that there is a dependency on state sync. Without state sync, while all nodes track all shards, our job would be easier actually.